lib/two_factor_authentication/models/two_factor_authenticatable.rb

```diff
@@ -11,17 +11,37 @@ module ClassMethods
         def has_one_time_password(options = {})
           include InstanceMethodsOnActivation
           include EncryptionInstanceMethods if options[:encrypted] == true
-
-          before_create { populate_otp_column }
         end
 
         ::Devise::Models.config(
-          self, :max_login_attempts, :allowed_otp_drift_seconds, :otp_length,
-          :remember_otp_session_for_seconds, :otp_secret_encryption_key)
+          self, :max_login_attempts, :allowed_otp_drift_seconds, :otp_length, :direct_otp_length,
+          :direct_otp_valid_for, :remember_otp_session_for_seconds, :otp_secret_encryption_key)
       end
 
       module InstanceMethodsOnActivation
         def authenticate_otp(code, options = {})
+          return true if direct_otp and authenticate_direct_otp(code, options)
+          return true if totp_enabled? and authenticate_totp(code, options)
+          false
+        end
+
+        def authenticate_direct_otp(code, options = {})
+          return false if direct_otp.nil?
+
+          return false if direct_otp != code
+
+          if Time.now.utc > direct_otp_sent_at + self.class.direct_otp_valid_for
+            # OTP has expired
+            return false
+          end
+
+          # OTP is valid, but can only be used once so clear it from the DB
+          update_columns(direct_otp: nil, direct_otp_sent_at: nil)
+          true
+        end
+
+        def authenticate_totp(code, options = {})
+          raise "authenticate_totp called with no otp_secret_key set" if otp_secret_key.nil?
```

Comment:
I don't think this is needed because the only time `authenticate_totp` will be called is via the `authenticate_otp` method in this line:
```
return true if totp_enabled? and authenticate_totp(code, options)
```
If `totp_enabled?` returns false, `authenticate_totp` will never get called, so it's impossible for it to get called if otp_secret_key is nil